[Doc] Graph Construction PR Refactor - PR2: GPartition Doc (#918)
awslabs/graphstorm: docs/source/graph-construction/gs-processing/example.rst

.. _distributed_construction_example:

GraphStorm Distributed Construction Example
===========================================

> 

Author: We would like to put all the overview step in the index file. We provide an overview in https://github.com/awslabs/graphstorm/blob/39875c4954d540b9dfba61dda319623bcfd1a85e/docs/source/graph-construction/gs-processing/gspartition/index.rst. We also include the original dgl doc link there.

To demonstrate how to use distributed graph construction locally we will
use the same example data as we use in our
unit tests, which you can find in the project's repository,
under ``graphstorm/graphstorm-processing/tests/resources/small_heterogeneous_graph``.

Install example dependencies
----------------------------

To run the local example you will need to install the GSProcessing and GraphStorm
library to your Python environment, and you'll need to clone the

> GraphStorm repository to get access to the data, and dgl tool for GPartition.

Reviewer: ```suggestion
GraphStorm repository to get access to the data, and dgl tool for GSPartition.
```

Follow the :ref:`gsp-installation-ref` guide to install the GSProcessing library.

> To install the GPartition library:

Author: Both GraphStorm Dependencies + DGL tool script + pydantic.

> To install the GPartition library:

Reviewer 2: Let's change the naming then, gpartition is a module of GSF (with its own entry points), not a separate library.

> To install the GPartition library:

Author: I change the description here.

> To install the GPartition library:

Reviewer: ```suggestion
To install the GSPartition library:
```

.. code-block:: bash

    pip install graphstorm
    pip install pydantic
    pip install torch==2.1.0 --index-url https://download.pytorch.org/whl/cpu
    pip install dgl==1.1.3 -f https://data.dgl.ai/wheels-internal/repo.html
    git clone https://github.com/awslabs/graphstorm.git
    cd graphstorm

>     git clone --branch v1.1.3 https://github.com/dmlc/dgl.git

Author: Not refer to the guide we have in the quick-start, because we need to install the tool script and pydantic, which are not required in the main gsf repo.

>     git clone --branch v1.1.3 https://github.com/dmlc/dgl.git

Author: We need dgl tool script for dispatch data.

You can then navigate to the ``graphstorm-processing/`` directory
that contains the relevant data:

.. code-block:: bash

    cd ./graphstorm-processing/

Expected file inputs and configuration
--------------------------------------

> The example will include GSProcessing as the first step and GPartition as the second step.

Author: I added one for the total distributed graph construction like what dgl did.

GSProcessing expects the input files to be in a specific format that will allow
us to perform the processing and prepare the data for partitioning and training.

> GPartition can handle the output by the GSProcessing.

Reviewer 2: ```suggestion
GPartition then takes the output of GSProcessing to produce graph files ready for training.
```

    for node id 1 etc.

> Run a GPartition job locally
> ------------------------------

Reviewer 2: ```suggestion
Run a GPartition job locally
---------------------------
```

> While :ref:`GPartition<gpartition_index>` is designed for distributed cluster,

Reviewer 2: ```suggestion
While :ref:`GPartition<gpartition_index>` is designed to run on a multi-machine cluster,
```

you can run GPartition job locally for the example. Once you have completed previous installation

> and GSProcessing example, you can proceed to run the GPartition example.

Author: It is the first section in the following paragraph.

Assuming your working directory is ``graphstorm``,
you can use the following command to run the processing job locally:

.. code:: bash

    echo 127.0.0.1 > ip_list.txt
    python3 -m graphstorm.gpartition.dist_partition_graph \
        --input-path /tmp/gsprocessing-example/ \
        --metadata-filename updated_row_counts_metadata.json \
        --output-path /tmp/gpartition-example/ \
        --num-parts 1 \
        --dgl-tool-path ./dgl/tools \
        --partition-algorithm random \
        --ip-config ip_list.txt \

>         --do-dispatch

Reviewer 2: For this case, do we still need the ip config for GSF init? I thought we removed that requirement?

Author: It should be necessary if we want to do the dispatch: https://github.com/awslabs/graphstorm/blob/main/python/graphstorm/gpartition/dist_partition_graph.py#L189

>         --do-dispatch

Reviewer 2: Comes from the original implementation https://github.com/awslabs/graphstorm/pull/548/files

I think you added it to allow us to perform partition assignment without the dispatch step, but agree the name is not meaningful to external users. We can rename in a separate PR to `--build-distributed-graph` or do the opposite and have something like `--partition-assignment-only`, where we always do the graph building, unless this arg is provided.

>         --do-dispatch

Author: I would prefer not to change the command's function here because it will lead to some break changes for the whole command logic.

> Examining the job output for GPartition
> ------------------------------------------

Reviewer 2: ```suggestion
Examining the job output of GPartition
--------------------------------------
```

> Once the partition job is done, you can examine the outputs they created.

Reviewer 2: ```suggestion
Once the partition job is done, you can examine the outputs is has created.
```

Author: change the context to "the outputs it has created".

.. code-block:: bash

    $ cd /tmp/gpartition-example

>     $ ls -l

Reviewer 2: If we want recursive output use
```suggestion
    $ ls -ltR
```

    dist_graph/
        metadata.json
        |- part0/
            edge_feat.dgl
            graph.dgl
            node_feat.dgl
            orig_eids.dgl
            orig_nids.dgl
    partition_assignment/
        director.txt
        genre.txt
        movie.txt
        partition_meta.json
        user.txt

> Now there are a few directories containing partitioned graph ready for training and inference.

Author: Mentioned part*/ under the dist_graph

* ``metadata.json``: This file contains metadata about the distributed DGL graph. The files under
``dist_graph`` should be ready for the training/inference pipeline.
* ``part0``: As we only specify 1 partition in the previous command, so we only have one part here.
There are five kinds of data here:
    * ``edge_feat.dgl``: The partitioned edge features for part 0 stored in binary format.
    * ``graph.dgl``: The graph structure data for part 0 stored in binary format.
    * ``node_feat.dgl``: The node features data for part 0 stored in binary format.
    * ``orig_eids.dgl``: The mapping for edges between raw edge IDs and graph's edge IDs.
    * ``orig_nids.dgl``: The mapping for nodes between raw node IDs and graph's node IDs.

> * ``partition_assignment``: This directory should contain different partition results for different node types,

Author: Here should be the only part we detailed introduce each file generated by the GPartition. We follow the same behavior as what we did in the GSProcessing.

> the results can reused for the : `dgl dispatch pipeline <https://docs.dgl.ai/en/latest/guide/distributed-preprocessing.html#distributed-graph-partitioning-pipeline>`_

Reviewer 2: Ensure the lists render correctly here, sometimes multi-line list items mess up the formatting.

Author: It should be. The related preview is here: https://jalencato-graphstorm-doc.readthedocs.io/en/gsprocessing-gpartition-doc/graph-construction/gs-processing/example.html#examining-the-job-output-of-gpartition

    for node id 1 etc.

> Run a GPartition job locally

Reviewer: ```suggestion
Run a GSPartition job locally
```

---------------------------
While :ref:`GPartition<gpartition_index>` is designed to run on a multi-machine cluster,

> you can run GPartition job locally for the example. Once you have completed previous installation

Reviewer: ```suggestion
you can run GPartition job locally for the example. Once you have completed the installation
```

> and GSProcessing example, you can proceed to run the GPartition example.

Reviewer: ```suggestion
and GSProcessing example described in the previous sections, you can proceed to run the GSPartition example.
```

Assuming your working directory is ``graphstorm``,

> you can use the following command to run the processing job locally:

Reviewer: ```suggestion
you can use the following command to run the partition job locally:
```

.. code:: bash

    echo 127.0.0.1 > ip_list.txt
    python3 -m graphstorm.gpartition.dist_partition_graph \
        --input-path /tmp/gsprocessing-example/ \
        --metadata-filename updated_row_counts_metadata.json \

>         --output-path /tmp/gpartition-example/ \

Reviewer: ```suggestion
        --output-path /tmp/gspartition-example/ \
```

        --num-parts 1 \
        --dgl-tool-path ./dgl/tools \
        --partition-algorithm random \
        --ip-config ip_list.txt \
        --do-dispatch

> Examining the job output of GPartition

Reviewer: ```suggestion
Examining the job output of GSPartition
```

--------------------------------------

Once the partition job is done, you can examine the outputs it has created.

.. code-block:: bash

>     $ cd /tmp/gpartition-example

Reviewer: ```suggestion
    $ cd /tmp/gspartition-example
```

    $ ls -ltR

    dist_graph/
        metadata.json
        |- part0/
            edge_feat.dgl
            graph.dgl
            node_feat.dgl
            orig_eids.dgl
            orig_nids.dgl
    partition_assignment/
        director.txt
        genre.txt
        movie.txt
        partition_meta.json
        user.txt

> Now there are a few directories containing partitioned graph ready for training and inference.

Reviewer: ```suggestion
There are a few directories containing partitioned graph ready for training and inference.
```

* ``metadata.json``: This file contains metadata about the distributed DGL graph. The files under

> ``dist_graph`` should be ready for the training/inference pipeline.

Reviewer: Should the ``dist_graph`` folder be introduced first before talking about the ``metadata.json``?

And,

May change to "The folder ``dist_graph`` contains files ready for GraphStorm training and inference, including:"

> * ``part0``: As we only specify 1 partition in the previous command, so we only have one part here.

Reviewer: ```suggestion
* ``part0``: As we only specify 1 partition in the previous command, so we only have one part folder here.
```

There are five kinds of data here:

>     * ``edge_feat.dgl``: The partitioned edge features for part 0 stored in binary format.

Reviewer: ```suggestion
    * ``edge_feat.dgl``: The edge features for part 0 stored in binary format.
```

    * ``graph.dgl``: The graph structure data for part 0 stored in binary format.
    * ``node_feat.dgl``: The node features data for part 0 stored in binary format.

>     * ``orig_eids.dgl``: The mapping for edges between raw edge IDs and graph's edge IDs.

Reviewer: ```suggestion
    * ``orig_eids.dgl``: The mapping for edges between raw edge IDs and the partitioned graph edge IDs.
```

>     * ``orig_nids.dgl``: The mapping for nodes between raw node IDs and graph's node IDs.

Reviewer: ```suggestion
    * ``orig_nids.dgl``: The mapping for nodes between raw node IDs and the partitioned graph node IDs.
```

> * ``partition_assignment``: This directory should contain different partition results for different node types,

Reviewer: ```suggestion
``partition_assignment``: This directory contains different partition results for different node types,
```

> the results can reused for the : `dgl dispatch pipeline <https://docs.dgl.ai/en/latest/guide/distributed-preprocessing.html#distributed-graph-partitioning-pipeline>`_

Reviewer: Broken sentences. The "the results can ..." seems have no relation with the previous sentence.

    for node id 1 etc.

Run a GSPartition job locally
------------------------------
While :ref:`GSPartition<gspartition_index>` is designed to run on a multi-machine cluster,

> you can run GSPartition job locally for the example. Once you have completed the installation
> and GSProcessing example described in the previous section, you can proceed to run the GSPartition.

Reviewer 2: ```suggestion
you can run GSPartition job locally for the example. Once you have completed the installation
and the GSProcessing example described in the previous section, you can proceed to run the GSPartition step.
```

Assuming your working directory is ``graphstorm``,
you can use the following command to run the partition job locally:

.. code:: bash

    echo 127.0.0.1 > ip_list.txt
    python3 -m graphstorm.gpartition.dist_partition_graph \
        --input-path /tmp/gsprocessing-example/ \
        --metadata-filename updated_row_counts_metadata.json \
        --output-path /tmp/gspartition-example/ \

>         --num-parts 1 \

Author: Test successfully with 2 here. Let us make it to 2.

        --dgl-tool-path ./dgl/tools \
        --partition-algorithm random \
        --ip-config ip_list.txt 

The command above will first do graph partitioning to determine the ownership for each partition and save the results.
Then it will do data dispatching to physically assign the partitions to graph data and dispatch them to each machine.
Finally it will generate the graph data ready for training/evaluation.

Examining the job output of GSPartition
---------------------------------------

Once the partition job is done, you can examine the outputs.

.. code-block:: bash

    $ cd /tmp/gspartition-example
    $ ls -ltR

    dist_graph/
        metadata.json
        |- part0/
            edge_feat.dgl
            graph.dgl
            node_feat.dgl
            orig_eids.dgl
            orig_nids.dgl
    partition_assignment/
        director.txt
        genre.txt
        movie.txt
        partition_meta.json
        user.txt

The ``dist_graph`` folder contains partitioned graph ready for training and inference.

> * ``part0``: As we only specify 1 partition in the previous command, we have one part folder here.

Reviewer 3: ```suggestion
``part0``: As we only specify 1 partition in the previous command, we have one part folder here.
```

Author: I think we still need * because it is the subfolder under the ```dist_graph```?

> There are five data here:

Reviewer 2: ```suggestion
There are five files for the partition
```

    * ``edge_feat.dgl``: The edge features for part 0 stored in binary format.
    * ``graph.dgl``: The graph structure data for part 0 stored in binary format.
    * ``node_feat.dgl``: The node features data for part 0 stored in binary format.
    * ``orig_eids.dgl``: The mapping for edges between raw edge IDs and the partitioned graph edge IDs.
    * ``orig_nids.dgl``: The mapping for nodes between raw node IDs and the partitioned graph node IDs.

> * ``metadata.json``: This file contains metadata about the distributed DGL graph.

Reviewer 3: ```suggestion
``metadata.json``: This file contains metadata about the distributed DGL graph.
```

Author: Ditto.
